### PR TITLE
Add support for impersonation for certificate requests

### LINKED ISSUE #331: Add support for Istio with Ambient mode
## Description

With the release of Istio's Ambient mode (Beta), decided to transition from the Istio Sidecars + Istio-CSR setup to Istio Ambient + Istio-CSR. Successfully redeployed Istio with Ambient mode (v1.22.0), but encountered an issue in Istio-CSR related to identity matching. The error occurs when identities are proxied via Ztunnel to Istio-CSR.

Istio Ambient mode Beta version was released and decided try to switch from **Istio Sidecars + Istio-CSR** to setup with **Istio Ambient + Istio-CSR**.

Re-deploy Istio with Ambient mode (v1.22.0) went successfully but issue pop out in istio-csr with identity match for spiffe proxied via Ztunnel to istio-csr. 

**Error in istio-csr pod:**
```
error	klog	[spiffe://cluster.local/ns/istio-system/sa/ztunnel] != [spiffe://cluster.local/ns/example/sa/my-app]: failed to match URIs with identities	logger=grpc-server serving-addr=0.0.0.0:6443 identities=spiffe://cluster.local/ns/istio-system/sa/ztunnel
```
**error in ztunnel pod:**
```
error	access	connection complete	src.addr=10.0.0.121:44562 src.workload=my-app-595cf9cf9c-db6x8 src.namespace=example src.identity="spiffe://cluster.local/ns/example/sa/my-app" dst.addr=10.0.0.58:15008 dst.hbone_addr=10.0.0.58:8080 dst.service=my-app.example.svc.cluster.local dst.workload=my-app-78f4dff8d8-fgfp7 dst.namespace=example dst.identity="spiffe://cluster.local/ns/example/sa/my-app" direction="outbound" bytes_sent=0 bytes_recv=0 duration="0ms" error="identity error: signing gRPC error (The request does not have valid authentication credentials): request authenticate failure
```

Reaching out Istio Ambient team on Slack confirmed that support for Istio-CSR in Ambient mode needs to be implemented in Istio-CSR.

## Used Helm values for Istio with sidecars before switch to Ambient mode

**cert-manager/istio-csr helm values:**
```yaml
app:
  tls:
    rootCAFile: /var/run/secrets/istio-csr/ca.pem
    certificateDNSNames:
      - istio-csr.cert-manager.svc
      - cert-manager-istio-csr.cert-manager.svc

volumes:
- name: root-ca
  secret:
    secretName: istio-root-ca

volumeMounts:
- name: root-ca
  mountPath: /var/run/secrets/istio-csr
```
**istio/istiod helm values:**
```yaml
global:
  caAddress: cert-manager-istio-csr.cert-manager.svc:443

pilot:
  extraContainerArgs:
    - --tlsCertFile=/etc/cert-manager/tls/tls.crt
    - --tlsKeyFile=/etc/cert-manager/tls/tls.key
    - --caCertFile=/etc/cert-manager/ca/root-cert.pem

  env: 
    ENABLE_CA_SERVER: false

  volumeMounts:
    - mountPath: /etc/cert-manager/tls
      name: cert-manager
      readOnly: true
    - mountPath: /etc/cert-manager/ca
      name: istio-csr-ca-configmap
      readOnly: true

  # Additional volumes to the istiod pod
  volumes: []
    - name: cert-manager
      secret:
        defaultMode: 420
        secretName: istiod-tls
    - configMap:
        defaultMode: 420
        name: istio-ca-root-cert
        optional: true
      name: ca-root-cert
```
## Changes and switch to Ambient mode and Ztunnel

I followed the instructions for installing Ambient Mesh as outlined here: [Istio Ambient Mesh Installation](https://istio.io/latest/docs/ambient/install/helm-installation/). The CA Address was configured in the same manner as the previous sidecar setup.

For ztunnel set that CA Address will be Istio-CSR (same address was used with Istio sidecars):

**istio/ztunnel helm values:**
``` yaml
caAddress: "cert-manager-istio-csr.cert-manager.svc:443"
```

## Note

Switching CA from Istio-CSR to the default Istiod ambient mode setup without Istio-CSR configuration our ambient mesh started working.

### PR DESCRIPTION
**Pull request by paulwilljones**
Resolves #331

### CODE REVIEW

File: pkg/server/auth.go
Comment (howardjohn) [inline pkg/server/auth.go:107]:
```suggestion
		log.Error(fmt.Errorf("%v != %v", s.impersonatedIdentity, csr.URIs), "failed to match URIs with impersonated identities")
```

nit: might be nice to help distinguish

**Review by howardjohn — COMMENTED** [review]
:+1:

File: pkg/server/node_auth.go
Comment (inteon) [inline pkg/server/node_auth.go:17]: @paulwilljones does Istio have a similar ClusterNodeAuthorizer in their own implementation?

Comment (paulwilljones) [reply pkg/server/node_auth.go:17]: Yes, [`ClusterNodeAuthorizer`](https://github.com/istio/istio/blob/1.22.1/security/pkg/server/ca/node_auth.go#L74) is the same, however, I've reimplemented it here to make sure the pod informer is synced before the index is created.

Comment (inteon) [reply pkg/server/node_auth.go:17]: Could you add a link to that implementation in this file?

**Comment by inteon** [comment]
@paulwilljones the automated tests found some issues: https://storage.googleapis.com/cert-manager-prow-artifacts/pr-logs/pull/cert-manager_istio-csr/336/pull-cert-manager-istio-csr-verify/1813889849741021184/build-log.txt

**Comment by paulwilljones** [comment]
@inteon @SgtCoDFish I added an e2e test for Ambient. Can we add this to Prow for Istio 1.22 onwards?

**Comment by inteon** [comment]
> @inteon @SgtCoDFish I added an e2e test for Ambient. Can we add this to Prow for Istio 1.22 onwards?

The tests are defined here: https://github.com/cert-manager/testing/blob/master/config/jobs/cert-manager/istio-csr/cert-manager-istio-csr-presubmits.yaml
We probably should start by adding this as an optional test so other PRs are not blocked.

**Comment by paulwilljones** [comment]
> > @inteon @SgtCoDFish I added an e2e test for Ambient. Can we add this to Prow for Istio 1.22 onwards?
> 
> The tests are defined here: https://github.com/cert-manager/testing/blob/master/config/jobs/cert-manager/istio-csr/cert-manager-istio-csr-presubmits.yaml We probably should start by adding this as an optional test so other PRs are not blocked.

Added https://github.com/cert-manager/testing/pull/1050 to run Ambient e2e tests for 1.22

File: pkg/server/node_auth.go
Comment (SgtCoDFish) [inline pkg/server/node_auth.go:93]: turbo nitpick (not a blocker): There are a few comments here that have `// //`, which I guess is an artifact from copy-pasting?

File: pkg/server/node_auth.go
Comment (SgtCoDFish) [inline pkg/server/node_auth.go:99]: here too!

File: test/e2e/suite/runtimeconfiguration/runtimeconfiguration.go
Comment (SgtCoDFish) [inline test/e2e/suite/runtimeconfiguration/runtimeconfiguration.go:94]: question: are you intending to do this before merging this PR?

**Review by SgtCoDFish — CHANGES_REQUESTED** [review]
With no exaggeration this is one of the best PRs I've been able to review! Thank you!

I've got a few little bits here and there and a few questions about how to proceed with this, but I don't see any major blockers!

**Comment by SgtCoDFish** [comment]
It looks like #357 caused quite a few conflicts with this PR - sorry about that!

File: test/e2e/suite/runtimeconfiguration/runtimeconfiguration.go
Comment (paulwilljones) [reply test/e2e/suite/runtimeconfiguration/runtimeconfiguration.go:94]: this currently fails for ambient as the `istioctl proxy-config` command to retrieve the certificates is not valid for ztunnel, and `istioctl experimental ztunnel-config certificates` doesn't return the required output. We can block the PR on this until there's a workaround or fix once `istioctl` moves out of experimental.

File: go.mod
Comment (paulwilljones) [reply go.mod]: as discussed, `go mod tidy` sets these versions and compatibility between `istio.io/api` and `istio.io/istio` is difficult to correlate.

File: deploy/charts/istio-csr/values.yaml
Comment (inteon) [inline deploy/charts/istio-csr/values.yaml:209]: should we default this to `istio-system/ztunnel`?

Comment (inteon) [reply deploy/charts/istio-csr/values.yaml:209]: in which situation do we want to change this value?

Comment (howardjohn) [reply deploy/charts/istio-csr/values.yaml:209]: Users may run ztunnel under a different service account. For example https://istio.io/latest/docs/ambient/install/platform-prerequisites/#google-kubernetes-engine-gke

**Comment by inteon** [comment]
I'll upgrade the istio package in https://github.com/cert-manager/istio-csr/pull/385, this should reduce the diff for this PR a bit.

File: deploy/charts/istio-csr/values.yaml
Comment (SgtCoDFish) [inline deploy/charts/istio-csr/values.yaml:209]: nitpick (non-blocking): It might be worth saying that this is a comma-separated list, to make it easier for people to know how to specify the values.

```suggestion
    # A comma-separated list of service accounts that are allowed to use node authentication for CSRs, eg. "istio-system/ztunnel"
    caTrustedNodeAccounts: ""
```

File: make/test-e2e.mk
Comment (SgtCoDFish) [inline make/test-e2e.mk:75]: suggestion: There's a typo in `ISTIO_INSTALL_OPIONS` - should be `ISTIO_INSTALL_OPTIONS`.

File: test/e2e/suite/mtls/mtls.go
Comment (SgtCoDFish) [inline test/e2e/suite/mtls/mtls.go:67]: question: Maybe an obvious question (and not really related to the PR), but just checking I've understood right: I guess there's no situation where you'd set dataplane-mode to ambient and enable injection? I can't imagine why that would be desirable, just want to check.

**Review by SgtCoDFish — CHANGES_REQUESTED** [review]
Got a few comments on this but nothing huge. Feels like it's really close!

File: test/e2e/suite/mtls/mtls.go
Comment (SgtCoDFish) [reply test/e2e/suite/mtls/mtls.go:67]: Thanks, that's really helpful 👍

**Review by SgtCoDFish — APPROVED** [review]
/lgtm
/approve

Thanks for all your work on this, it's really cool and I'm hoping we'll get to release this out for people to play with soon 🚀

**Pull request merged**
